sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitwiseExpressions.scala

```diff
@@ -204,3 +205,48 @@ case class BitwiseCount(child: Expression)
     case LongType => java.lang.Long.bitCount(input.asInstanceOf[Long])
   }
 }
+
+object BitwiseGet {
+  def calculate(target: Int, pos: Int): Int = {
+    val bitCount = java.lang.Long.bitCount(target)
```

Comment:
This is the number of set bits:
```     
    * Returns the number of one-bits in the two's complement binary
     * representation of the specified {@code long} value.  This function is
     * sometimes referred to as the <i>population count</i>.
```
The comparison should be against the `java.lang.Long.SIZE` or for each datatype the constant should be the one defined in the corresponding class (Integer, Short, Byte).  This is another reason to use `IntegralType`.

Reply:
Why we need compare against the` java.lang.Long.SIZE`? we only need the number of bit in the integer value.

Reply:
because `checkEvaluation(BitwiseGet(Literal(11), Literal(4)), 0)` will fail with `Cause: java.lang.IllegalArgumentException: Invalid bit position: 4 exceeds the bit length of the target integer`.
For integer the valid inputs for second parameter should be 0 to 31. currently you can't check past the most significant set bit.

And for each different input type the range should be different (for long 0..63, etc).
The number of bits are kept in the constants `java.lang.Long/Integer/Short/Byte.SIZE`

Reply:
The second parameter just cannot be negative.

Reply:
I never said, that it should be. The problem is with the upper bound.

In the failing example the second parameter is 4.
Add `checkEvaluation(BitwiseGet(Literal(11), Literal(4)), 0)` to the test,

Reply:
This test will throw
`Caused by: java.lang.IllegalArgumentException: Invalid bit position: 4 exceeds the bit length of the target integer` as my expected.

Reply:
In fact, this test case is similar to the one https://github.com/apache/spark/blob/bcc51357bb9a62dde040923986334efcf2ac8c14/sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala#L192

Reply:
Yes, but it should return 0, because the 5th bit in 11 is 0.

Reply:
Even though 11 has set bits among the 4 lowest bits, I should still be able to access the higher bits. 
The range is defined by the data type (0..63 for Long, 0..31 for Integer, etc) not by the actual value.

Reply:
The document of teradata tells us: `If target_bit_arg is negative or out-of-range (meaning that it exceeds the size of target_arg), an error is returned.`

Reply:
The behavior of impala is: 
```
select getbit(-1,25); /* 11111111 */
ERROR: Invalid bit position: 25
```

Reply:
Well if the input is a byte (given the pattern in the comment), then yes 25 is out of range, because a byte has 8 bits.
what does this return: `getbit(0, 3)`?
I would assume that this is 0, but your implementation would throw an exception.

Reply:
Impala's behavior looks a little strange. As you said, getbit(0, 3) in Snowflake returns 0.

Reply:
You comment on https://github.com/apache/spark/pull/31170#discussion_r557287080 looks better. I will use java.lang.Long.SIZE not the fact size.